llvm/lib/Target/AMDGPU/AMDGPUNextUseAnalysis.cpp

```diff
@@ -0,0 +1,573 @@
+//===---------------------- AMDGPUNextUseAnalysis.cpp  --------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "AMDGPUNextUseAnalysis.h"
+#include "AMDGPU.h"
+#include "GCNSubtarget.h"
+#include "llvm/InitializePasses.h"
+#include <limits>
+#include <queue>
+
+using namespace llvm;
+
+#define DEBUG_TYPE "amdgpu-next-use-analysis"
+
+static cl::opt<bool> DumpNextUseDistance("dump-next-use-distance",
+                                         cl::init(false), cl::Hidden);
+
+bool AMDGPUNextUseAnalysis::isBackedge(MachineBasicBlock *From,
+                                       MachineBasicBlock *To) const {
+  MachineLoop *Loop1 = MLI->getLoopFor(From);
+  MachineLoop *Loop2 = MLI->getLoopFor(To);
+  if (!Loop1 || !Loop2 || Loop1 != Loop2)
+    return false;
+  SmallVector<MachineBasicBlock *, 2> Latches;
+  Loop1->getLoopLatches(Latches);
+  auto It = llvm::find(Latches, From);
+  MachineBasicBlock *LoopHeader = Loop1->getHeader();
+  return It != Latches.end() && From->isSuccessor(To) && To == LoopHeader;
+}
+
+// Calculate the shortest distance between two blocks using Dijkstra algorithm.
+std::pair<SmallVector<MachineBasicBlock *>, uint64_t>
+AMDGPUNextUseAnalysis::getShortestPath(MachineBasicBlock *FromMBB,
+                                       MachineBasicBlock *ToMBB) {
+  assert(FromMBB != ToMBB && "The basic blocks should be different.\n");
+  DenseSet<MachineBasicBlock *> Visited;
+  struct Data {
+    MachineBasicBlock *BestPred = nullptr;
+    uint64_t ShortestDistance = std::numeric_limits<uint64_t>::max();
+  };
+  DenseMap<MachineBasicBlock *, Data> MBBData;
+
+  auto Cmp = [&MBBData](MachineBasicBlock *MBB1, MachineBasicBlock *MBB2) {
+    return MBBData[MBB1].ShortestDistance > MBBData[MBB2].ShortestDistance;
+  };
+  std::priority_queue<MachineBasicBlock *, std::vector<MachineBasicBlock *>,
+                      decltype(Cmp)>
+      Worklist(Cmp);
+
+  Worklist.push(FromMBB);
+  MBBData[FromMBB] = {nullptr, 0};
+
+  while (!Worklist.empty()) {
+    MachineBasicBlock *CurMBB = Worklist.top();
+    Worklist.pop();
+
+    if (!Visited.insert(CurMBB).second)
+      continue;
+
+    if (CurMBB == ToMBB) {
+      // We found the destination node, build the path ToMBB->...->FromMBB
+      SmallVector<MachineBasicBlock *> Path;
+      MachineBasicBlock *PathMBB = ToMBB;
+      while (PathMBB != nullptr) {
+        Path.push_back(PathMBB);
+        if (PathMBB == FromMBB)
+          break;
+        auto It = MBBData.find(PathMBB);
+        PathMBB = It != MBBData.end() ? It->second.BestPred : nullptr;
+      }
+      assert(Path.back() == FromMBB && "Incomplete path!");
+      auto *Pred = MBBData[CurMBB].BestPred;
```

Comment:
At this point, `CurMBB` should be equal to `ToMBB` right ?
```suggestion
      auto *Pred = MBBData[ToMBB].BestPred;
```